### PR TITLE
Fix deployment agent failure when docker images cannot be fetched

### PR DESCRIPTION
**Pull request by mrnicegyu11**
## What do these changes do?

Fix a bug in the deployment agent, due to a type mismatch. This bug may occur when fetching docker images fails at one point in time, and then works again.

The most noticeable impact of this bug is that - if a mattermost client is configured - there will be erroneous mattermost notifications sent out every 3-5 minutes.

Bonus: Add test

Help with pytest by @GitHK is gratefully acknowledged.

## Checklist

- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes

### CODE REVIEW

File: tests/unit/test_docker_registries_watcher.py [hunk 1]
```diff
@@ -89,3 +89,24 @@ async def test_docker_registries_watcher(
         "ubuntu": "image signature changed",
     }
     _assert_docker_client_calls(mock_docker_client, registry_config, valid_docker_stack)
+
```
Comment (GitHK) [inline tests/unit/test_docker_registries_watcher.py:92]: I would create two new tests out of all the below

File: tests/unit/test_docker_registries_watcher.py [hunk 2]
```diff
@@ -89,3 +89,24 @@ async def test_docker_registries_watcher(
...
+    # Handle the failure of fetching an image
+    mock_docker_client.return_value.images.get_registry_data.return_value.attrs = {
+        "Descriptor": "somenewsignature2"
+    }
+    mock_docker_client.return_value.images.get_registry_data.side_effect = (
+        docker.errors.APIError("Mocked Error Image cant be fetched")
+    )
```
Comment (GitHK) [inline tests/unit/test_docker_registries_watcher.py:99]: I would put all this in a fixture

File: tests/unit/test_docker_registries_watcher.py [hunk 3]
```diff
@@ -89,3 +89,24 @@ async def test_docker_registries_watcher(
...
+    change_result = await docker_watcher.check_for_changes()
+    assert change_result == {}
+    _assert_docker_client_calls(mock_docker_client, registry_config, valid_docker_stack)
+    mock_docker_client.return_value.images.get_registry_data.return_value.attrs = {
+        "Descriptor": "somenewsignature3"
+    }
+    mock_docker_client.return_value.images.get_registry_data.side_effect = None
```
Comment (GitHK) [inline tests/unit/test_docker_registries_watcher.py:106]: here you would have a second fixture

File: tests/unit/test_docker_registries_watcher.py [hunk 4]
```diff
@@ -89,3 +89,24 @@ async def test_docker_registries_watcher(
...
+    change_result = await docker_watcher.check_for_changes()
+    assert change_result == {
+        "jenkins:latest": "image signature changed",
+        "ubuntu": "image signature changed",
+    }
+    _assert_docker_client_calls(mock_docker_client, registry_config, valid_docker_stack)
```
Comment (GitHK) [inline tests/unit/test_docker_registries_watcher.py:112]: and a second test

File: tests/unit/test_docker_registries_watcher.py [hunk 5]
```diff
@@ -89,3 +89,24 @@ async def test_docker_registries_watcher(
...
+    # Handle the failure of fetching an image
+    mock_docker_client.return_value.images.get_registry_data.return_value.attrs = {
+        "Descriptor": "somenewsignature2"
+    }
+    mock_docker_client.return_value.images.get_registry_data.side_effect = (
+        docker.errors.APIError("Mocked Error Image cant be fetched")
+    )
+    change_result = await docker_watcher.check_for_changes()
+    assert change_result == {}
+    _assert_docker_client_calls(mock_docker_client, registry_config, valid_docker_stack)
```
Comment (GitHK) [inline tests/unit/test_docker_registries_watcher.py:102]: this would be your first test

**Review by GitHK — CHANGES_REQUESTED** [review]
Very nice overall.
Let's make it a bit better. Pleas split it up into two separate tests with appropriate names for what you are testing.
Advantages:
- simpler to read understand your intention
- simpler to debug and maintain

File: tests/unit/test_docker_registries_watcher.py [hunk 6]
```diff
@@ -89,3 +89,24 @@ async def test_docker_registries_watcher(
         "ubuntu": "image signature changed",
     }
     _assert_docker_client_calls(mock_docker_client, registry_config, valid_docker_stack)
+
+    # Handle the failure of fetching an image
```
Comment (pcrespov) [inline tests/unit/test_docker_registries_watcher.py:93]: very nice test!
Just a small suggestion:  i would create an independent test for this e.g. ``test_handle_failure_of_fetching_an_image``. It much better to have a lot of small unit tests with a very clear target than having one big-fat with a lot of checks and different contexts. If there are common parts, you can create helper functions or use the fixture.
Also mockups are typically part of the test setup, which is included in the fixture although when the test is small it is also acceptable to do it in the test-function body

**Review by pcrespov — APPROVED** [review]
Nice fix and even better test! Just some small suggestion about the test

**Review by sanderegg — APPROVED** [review]
great, I also agree that it would be nicer to have separate tests for separate issues (you could even put a link to the issue in the test)

**Review by GitHK — APPROVED** [review]
Looking better, thank!

**Pull request merged**